File: src/podcast/worker.py
"""
Background worker that polls the jobs table and executes pipeline steps.

Run with: python -m podcast.worker
"""

import asyncio
import json
import logging
import signal
from datetime import datetime, timezone

from sqlalchemy import case, select

from podcast.database import get_session
from podcast.log_handler import setup_logging, start_flush_loop, stop_flush_loop
from podcast.models import Episode, Job
from podcast.services.encoder import encode_mp3
from podcast.services.research import run_research
from podcast.services.transcript import generate_transcript
from podcast.services.tts import synthesize_speech

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)-8s %(name)s: %(message)s",
)
setup_logging("worker")
logger = logging.getLogger(__name__)

STEP_HANDLERS = {
    "research": run_research,
    "transcript": generate_transcript,
    "tts": synthesize_speech,
    "encode": encode_mp3,
}

NEXT_STEP = {
    "research": "transcript",
    "transcript": "tts",
    "tts": "encode",
    "encode": None,
}

EPISODE_STATUS_MAP = {
    "research": "researching",
    "transcript": "writing_transcript",
    "tts": "generating_audio",
    "encode": "encoding",
}

# Prioritise jobs closer to completion so we finish episodes rather than
# starting new ones.  Lower number = picked first.
STEP_PRIORITY = {
    "encode": 0,
    "tts": 1,
    "transcript": 2,
    "research": 3,
}

POLL_INTERVAL = 10  # seconds

_shutdown = False


def _handle_signal(signum, frame):
    global _shutdown
    logger.info("Received signal %s, shutting down gracefully...", signum)
    _shutdown = True


async def process_job(job_id, episode_id, step):
    """Execute a single job step. Returns metrics dict or None."""
    handler = STEP_HANDLERS.get(step)
    if not handler:
        raise ValueError(f"Unknown step: {step}")

    logger.info("Processing job %s: step=%s, episode=%s", job_id, step, episode_id)
    result = await handler(episode_id)
    return result if isinstance(result, dict) else None


async def _recover_stale_jobs():
    """Reset any jobs left in 'running' state from a previous worker that was killed.

    This handles the case where a deployment (SIGKILL after grace period) or
    crash interrupted a job mid-execution.  The job is set back to 'pending' so
    the normal poll loop picks it up.  TTS already has segment-level resume
    support, so re-running is efficient.
    """
    async with get_session() as db:
        result = await db.execute(
            select(Job).where(Job.status == "running")
        )
        stale_jobs = result.scalars().all()

        for job in stale_jobs:
            logger.warning(
                "Recovering stale job %s (step=%s, episode=%s) — resetting to pending",
                job.id,
                job.step,
                job.episode_id,
            )
            job.status = "pending"
            job.started_at = None

            # Reset the episode status so the UI doesn't show a stale state
            episode = await db.get(Episode, job.episode_id)
            if episode and episode.status != "failed":
                episode.status = EPISODE_STATUS_MAP.get(job.step, episode.status)

        if stale_jobs:
            logger.info("Recovered %d stale job(s)", len(stale_jobs))


async def poll_loop():
    """Main loop: pick up pending jobs and process them."""
    logger.info("Worker started, polling every %ds", POLL_INTERVAL)
    await start_flush_loop()

    try:
        await _recover_stale_jobs()
        await _poll_jobs()
    finally:
        await stop_flush_loop()


async def _poll_jobs():
    """Inner poll loop, separated so flush lifecycle wraps it."""
    while not _shutdown:
        try:
            # Pick up the next pending job
            job_id = None
            episode_id = None
            step = None

            async with get_session() as db:
                # Prefer jobs closer to completion (encode > tts >
                # transcript > research), then oldest first.
                step_priority = case(
                    *((Job.step == s, prio) for s, prio in STEP_PRIORITY.items()),
                    else_=99,
                )
                result = await db.execute(
                    select(Job)
                    .where(Job.status == "pending")
                    .order_by(step_priority, Job.created_at)
                    .limit(1)
                    .with_for_update(skip_locked=True)
                )
                job = result.scalar_one_or_none()

                if job is None:
                    await asyncio.sleep(POLL_INTERVAL)
                    continue

                # Capture IDs before session closes
                job_id = job.id
                episode_id = job.episode_id
                step = job.step

                # Mark as running
                job.status = "running"
                job.started_at = datetime.now(timezone.utc)
                job.attempts += 1

                # Update episode status
                episode = await db.get(Episode, episode_id)
                if episode:
                    episode.status = EPISODE_STATUS_MAP.get(step, episode.status)

            # Process outside the DB session to avoid long-held transactions
            try:
                metrics = await process_job(job_id, episode_id, step)

                # Mark job complete
                async with get_session() as db:
                    job_record = await db.get(Job, job_id)
                    job_record.status = "completed"
                    job_record.completed_at = datetime.now(timezone.utc)
                    if metrics:
                        job_record.metrics_json = json.dumps(metrics)

                    # Enqueue next step or mark as ready
                    next_step = NEXT_STEP.get(step)
                    episode = await db.get(Episode, episode_id)

                    if next_step:
                        new_job = Job(
                            episode_id=episode_id,
                            step=next_step,
                            status="pending",
                        )
                        db.add(new_job)
                        logger.info("Enqueued next step: %s for episode %s", next_step, episode_id)
                    else:
                        # Pipeline complete
                        episode.status = "ready"
                        episode.published_at = datetime.now(timezone.utc)
                        logger.info("Episode %s is ready!", episode_id)

            except Exception as e:
                logger.exception("Job %s failed: %s", job_id, e)
                async with get_session() as db:
                    job_record = await db.get(Job, job_id)
                    job_record.status = "failed"
                    job_record.error_message = str(e)

                    episode = await db.get(Episode, episode_id)
                    episode.status = "failed"
                    episode.error_message = str(e)
                    episode.failed_step = step

        except Exception as e:
            logger.exception("Unexpected error in poll loop: %s", e)
            await asyncio.sleep(POLL_INTERVAL)

    logger.info("Worker shut down.")


def main():
    signal.signal(signal.SIGTERM, _handle_signal)
    signal.signal(signal.SIGINT, _handle_signal)
    asyncio.run(poll_loop())


if __name__ == "__main__":
    main()

File: src/worker.ts
import { eq, sql, and } from "drizzle-orm";
import { db } from "./database.js";
import { closePool } from "./database.js";
import { episodes, jobs } from "./schema.js";
import { settings } from "./config.js";
import {
  setupLogging,
  startFlushLoop,
  stopFlushLoop,
  createLogger,
} from "./log-handler.js";
import { runResearch } from "./services/research.js";
import { generateTranscript } from "./services/transcript.js";
import { synthesizeSpeech } from "./services/tts.js";
import { encodeMp3 } from "./services/encoder.js";
import { getAllModelPricing } from "./services/llm-providers.js";

setupLogging("worker");
const log = createLogger("worker");

type StepHandler = (episodeId: string) => Promise<Record<string, unknown>>;

const STEP_HANDLERS: Record<string, StepHandler> = {
  research: runResearch,
  transcript: generateTranscript,
  tts: synthesizeSpeech,
  encode: encodeMp3,
};

const NEXT_STEP: Record<string, string | null> = {
  research: "transcript",
  transcript: "tts",
  tts: "encode",
  encode: null,
};

const EPISODE_STATUS_MAP: Record<string, string> = {
  research: "researching",
  transcript: "writing_transcript",
  tts: "generating_audio",
  encode: "encoding",
};

const STEP_PRIORITY: Record<string, number> = {
  encode: 0,
  tts: 1,
  transcript: 2,
  research: 3,
};

const POLL_INTERVAL = 10_000;
const COSTLY_STEPS = new Set(["research", "transcript", "tts"]);

const MODEL_PRICING = getAllModelPricing();
const DEFAULT_PRICING = { input: 3.0, output: 15.0 };

let shutdown = false;

async function todaysSpend(): Promise<number> {
  const todayStart = new Date();
  todayStart.setUTCHours(0, 0, 0, 0);

  const completedJobs = await db
    .select()
    .from(jobs)
    .where(
      and(
        eq(jobs.status, "completed"),
        sql`${jobs.completedAt} >= ${todayStart}`
      )
    );

  let total = 0;
  for (const job of completedJobs) {
    if (!job.metricsJson) continue;
    const metrics = JSON.parse(job.metricsJson);
    const inputT = metrics.input_tokens ?? 0;
    const outputT = metrics.output_tokens ?? 0;
    const model = metrics.model ?? "";
    const pricing = MODEL_PRICING[model] ?? DEFAULT_PRICING;
    total +=
      (inputT * pricing.input + outputT * pricing.output) / 1_000_000;
  }
  return total;
}

async function processJob(
  jobId: string,
  episodeId: string,
  step: string
): Promise<Record<string, unknown> | null> {
  const handler = STEP_HANDLERS[step];
  if (!handler) throw new Error(`Unknown step: ${step}`);

  log.info("Processing job %s: step=%s, episode=%s", jobId, step, episodeId);
  const result = await handler(episodeId);
  return typeof result === "object" ? result : null;
}

async function recoverStaleJobs(): Promise<void> {
  const staleJobs = await db
    .select()
    .from(jobs)
    .where(eq(jobs.status, "running"));

  for (const job of staleJobs) {
    log.warn(
      "Recovering stale job %s (step=%s, episode=%s) — resetting to pending",
      job.id,
      job.step,
      job.episodeId
    );
    await db
      .update(jobs)
      .set({ status: "pending", startedAt: null })
      .where(eq(jobs.id, job.id));

    // Reset the episode status
    const [episode] = await db
      .select()
      .from(episodes)
      .where(eq(episodes.id, job.episodeId))
      .limit(1);
    if (episode && episode.status !== "failed") {
      await db
        .update(episodes)
        .set({
          status: EPISODE_STATUS_MAP[job.step] ?? episode.status,
        })
        .where(eq(episodes.id, job.episodeId));
    }
  }

  if (staleJobs.length > 0) {
    log.info("Recovered %d stale job(s)", staleJobs.length);
  }
}

async function pollJobs(): Promise<void> {
  while (!shutdown) {
    try {
      // Atomically claim the next pending job (SELECT FOR UPDATE + mark running in one tx)
      const claimed = await db.transaction(async (tx) => {
        const result = await tx.execute(sql`
          SELECT id, episode_id, step
          FROM jobs
          WHERE status = 'pending'
          ORDER BY
            CASE step
              WHEN 'encode' THEN 0
              WHEN 'tts' THEN 1
              WHEN 'transcript' THEN 2
              WHEN 'research' THEN 3
              ELSE 99
            END,
            created_at
          LIMIT 1
          FOR UPDATE SKIP LOCKED
        `);

        const rows = result.rows as Array<{
          id: string;
          episode_id: string;
          step: string;
        }>;
        if (!rows || rows.length === 0) return null;

        const { id, episode_id: episodeId, step } = rows[0];

        // Gate costly steps behind daily spend limit
        if (COSTLY_STEPS.has(step) && settings.dailySpendLimit > 0) {
          const spend = await todaysSpend();
          if (spend >= settings.dailySpendLimit) {
            log.warn(
              "Daily spend limit ($%.2f/$%.2f) reached — skipping %s job %s",
              spend,
              settings.dailySpendLimit,
              step,
              id
            );
            return null;
          }
        }

        // Mark as running
        await tx
          .update(jobs)
          .set({
            status: "running",
            startedAt: new Date(),
            attempts: sql`${jobs.attempts} + 1`,
          })
          .where(eq(jobs.id, id));

        return { id, episodeId, step };
      });

      if (!claimed) {
        await sleep(POLL_INTERVAL);
        continue;
      }

      const { id: jobId, episodeId, step } = claimed;

      // Update episode status
      await db
        .update(episodes)
        .set({ status: EPISODE_STATUS_MAP[step] ?? "pending" })
        .where(eq(episodes.id, episodeId));

      // Process the job
      try {
        const metrics = await processJob(jobId, episodeId, step);

        // Mark job complete
        await db
          .update(jobs)
          .set({
            status: "completed",
            completedAt: new Date(),
            metricsJson: metrics ? JSON.stringify(metrics) : null,
          })
          .where(eq(jobs.id, jobId));

        // Enqueue next step or mark as ready
        const nextStep = NEXT_STEP[step];
        if (nextStep) {
          await db.insert(jobs).values({
            episodeId,
            step: nextStep,
            status: "pending",
          });
          log.info(
            "Enqueued next step: %s for episode %s",
            nextStep,
            episodeId
          );
        } else {
          // Pipeline complete
          await db
            .update(episodes)
            .set({ status: "ready", publishedAt: new Date() })
            .where(eq(episodes.id, episodeId));
          log.info("Episode %s is ready!", episodeId);
        }
      } catch (e: any) {
        log.error("Job %s failed: %s", jobId, e.message);
        await db
          .update(jobs)
          .set({ status: "failed", errorMessage: String(e) })
          .where(eq(jobs.id, jobId));
        await db
          .update(episodes)
          .set({
            status: "failed",
            errorMessage: String(e),
            failedStep: step,
          })
          .where(eq(episodes.id, episodeId));
      }
    } catch (e: any) {
      log.error("Unexpected error in poll loop: %s", e.message);
      await sleep(POLL_INTERVAL);
    }
  }

  log.info("Worker shut down.");
}

async function main(): Promise<void> {
  log.info("Worker started, polling every %ds", POLL_INTERVAL / 1000);
  startFlushLoop();

  process.on("SIGTERM", () => {
    log.info("Received SIGTERM, shutting down gracefully...");
    shutdown = true;
  });
  process.on("SIGINT", () => {
    log.info("Received SIGINT, shutting down gracefully...");
    shutdown = true;
  });

  try {
    await recoverStaleJobs();
    await pollJobs();
  } finally {
    await stopFlushLoop();
    await closePool();
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch((err) => {
  console.error("Worker fatal error:", err);
  process.exit(1);
});
